hedera-node/src/main/java/com/hedera/services/store/tokens/unique/UniqueTokenStore.java

```diff
@@ -63,34 +76,117 @@ public UniqueTokenStore(final EntityIdSource ids,
+		var res = mintProvisional(txBody, creationTime);
+		if (!res.equals(OK)) {
+			clearProvisional();
+			return res;
+		}
+
+		res = commitProvisional();
+		if (!res.equals(OK)) {
+			clearProvisional();
+			return res;
+		}
+		clearProvisional();
+		return OK;
 	}
 
 	@Override
-	public ResponseCodeEnum mint(final TokenID tId, final String memo, final RichInstant creationTime) {
+	public ResponseCodeEnum commitProvisional() {
+		final AtomicBoolean didAnyFail = new AtomicBoolean(false);
+		if (!checkProvisional()) {
+			return INVALID_TRANSACTION_BODY;
+		} else {
+			provisionalUniqueTokens.forEach(provisionalTokenPair -> {
+				TokenID tokenId = provisionalTokenPair.getKey().tokenId().toGrpcTokenId();
+				String memo = provisionalTokenPair.getValue().getMemo();
+				RichInstant creationTime = provisionalTokenPair.getValue().getCreationTime();
+				var res = mintOne(tokenId, memo, creationTime);
+				// if native minting fails for some reason, revert changes
+				// serial num will not be incremented if we go here
+				if (!res.equals(OK)) {
+					var token = get(tokenId);
+					revertAdjustmentOfToken(token, tokenId);
+					// revert from map (remove)
+					uniqueTokenSupplier.get().remove(new MerkleUniqueTokenId(EntityId.fromGrpcTokenId(tokenId), token.getCurrentSerialNum()));
+					didAnyFail.set(true);
+				}
+			});
+			return didAnyFail.get() ? FAIL_INVALID : OK;
+		}
+	}
+
+	private ResponseCodeEnum mintOne(final TokenID tId, final String memo, final RichInstant creationTime) {
 		return tokenSanityCheck(tId, (merkleToken -> {
 			if (!merkleToken.hasSupplyKey()) {
 				return TOKEN_HAS_NO_SUPPLY_KEY;
 			}
-			var mintResult = super.mint(tId, 1);
-			if (!mintResult.equals(OK)) {
-				return mintResult;
+			var adjustmentResult = tryAdjustment(merkleToken.treasury().toGrpcAccountId(), tId, 1);
+			if (!adjustmentResult.equals(OK)) {
+				return adjustmentResult;
 			}
+			// Fails on this OP can happen only on lines above, so serial num should NOT be decremented
 			final var suppliedTokens = uniqueTokenSupplier.get();
 			final var eId = EntityId.fromGrpcTokenId(tId);
 			final var owner = merkleToken.treasury();
-			final long serialNum = merkleToken.incrementSerialNum();
-
+			final long serialNum = merkleToken.incrementSerialNum(); // increment only after all checks
 			final var nftId = new MerkleUniqueTokenId(eId, serialNum);
 			final var nft = new MerkleUniqueToken(owner, memo, creationTime);
 			suppliedTokens.put(nftId, nft);
+			lastMintedSerialNumbers.add(serialNum);
 			return OK;
 		}));
+	}
 
+	private void revertAdjustmentOfToken(MerkleToken token, TokenID tokenID) {
+		tryAdjustment(token.treasury().toGrpcAccountId(), tokenID, -1);
 	}
 
-	@Override
-	public ResponseCodeEnum wipe(final AccountID aId, final TokenID tId, final long wipingAmount, final boolean skipKeyCheck) {
-		return null;
+	private boolean checkProvisional() {
```

Comment:
Instead of this being performed before commit, we can move it in the `mint` method and fail once we prepared the `provisionalTokenSet`.